source/_posts/2018-01-31-infrastrucutre-as-code-a-beginners-perspective.markdown

```diff
@@ -0,0 +1,123 @@
+
+Some months go by and you need to add more queues and workers, but since you forgot to add that into a document it is not easy to do it the exact same way, you do it, it works, but now the queues and workers have slightly different configurations, this is called *Configuration drift*.
+
+You decide to add some levels of caching in you app you need Redis/Memcached, so you need to add more servers for those services. You also need to configure S3 to have you static files, a CDN to serve those worldwide, you need Route53 to configure your routes.
+
+By this time the company has grown, you need to create IAM roles and users for the new developers and give them the correct access rights.
+
+You need more reliability, so you need to create more servers in different datacenters, at this point you go nuts and close the company.
+
+The end of this last example was a bit catastrophic, but imagine yourself maintaining that amount of servers by hand, that does not scale up well.
+
+## Should you be doing IAC?
+
+As everything in tech, it depends. I personally can't imagine a project that would not be better by using IAC. So, in my opinion, if you have the chance to use it, give it a try and find out by yourself!
+
+Maybe if you are using Heroku for really simple apps it is too much overhead to have a tool like this, but if your Heroku project grows beyond a certain size you might benefit from this kind of tools as well. Take a look at [Terraform: Heroku Provider](https://www.terraform.io/docs/providers/heroku/index.html) for an example.
+
+# Terraform
+
+Terraform is a IAC tool from Hashicorp that solves the problems presented in the previous cases and more, and as the say in their site:
+
+> Terraform enables you to safely and predictably create, change, and improve infrastructure. It is an open source tool that codifies APIs into declarative configuration files that can be shared amongst team members, treated as code, edited, reviewed, and versioned.
+
+For example, with this simple code you have a load balancer and 5 VMs attached to it, and every time you run that code you will get to that same state, that is awesome! Imagine how many clicks you would have needed to do all that by hand, definitely a lot!
+
+{% codeblock lang:ruby %}
+resource "aws_elb" "frontend" {
+  name = "frontend-load-balancer"
+  listener {
+    instance_port     = 8000
+    instance_protocol = "http"
+    lb_port           = 80
+    lb_protocol       = "http"
+  }
+
+  instances = ["${aws_instance.app.*.id}"]
+}
+
+resource "aws_instance" "app" {
+  count         = 5
+  ami           = "ami-408c7f28"
+  instance_type = "t1.micro"
+}
+{% endcodeblock %}
+
+With code similar to this I was able to recreate the second case I presented in only a week, pretty amazing! I am sure it would have taken me at least double that time to do it manually.
```

Comment:
What do you think of *twice* instead of *double*?

Reply:
I prefer double hehe, according to the internet they are interchangeable in this case